rust/zypp-agama/zypp-agama-sys/c-layer/callbacks.cxx

```diff
@@ -390,39 +389,158 @@ struct KeyRingReport
 
 static KeyRingReport key_ring_report;
 
-struct DigestReceive : public zypp::callback::ReceiveReport<zypp::DigestReport> {
+struct DigestReceive
+    : public zypp::callback::ReceiveReport<zypp::DigestReport> {
   struct SecurityCallbacks *callbacks;
 
   DigestReceive() { callbacks = NULL; }
 
-  void set_callbacks(SecurityCallbacks *callbacks_) {
-    callbacks = callbacks_;
-  }
+  void set_callbacks(SecurityCallbacks *callbacks_) { callbacks = callbacks_; }
 
-  bool askUserToAcceptNoDigest( const zypp::Pathname &file ){
+  bool askUserToAcceptNoDigest(const zypp::Pathname &file) override {
     if (callbacks == NULL || callbacks->checksum_missing == NULL) {
       return zypp::DigestReport::askUserToAcceptNoDigest(file);
     }
-    return callbacks->checksum_missing(file.c_str(), callbacks->checksum_missing_data);
+    return callbacks->checksum_missing(file.c_str(),
+                                       callbacks->checksum_missing_data);
   }
 
-  bool askUserToAccepUnknownDigest( const zypp::Pathname &file, const std::string &name ) {
+  bool askUserToAccepUnknownDigest(const zypp::Pathname &file,
+                                   const std::string &name) override {
     if (callbacks == NULL || callbacks->checksum_unknown == NULL) {
       return zypp::DigestReport::askUserToAccepUnknownDigest(file, name);
     }
-    return callbacks->checksum_unknown(file.c_str(), name.c_str(), callbacks->checksum_unknown_data);
+    return callbacks->checksum_unknown(file.c_str(), name.c_str(),
+                                       callbacks->checksum_unknown_data);
   }
 
-  bool askUserToAcceptWrongDigest( const zypp::Pathname &file, const std::string &requested, const std::string &found ) {
+  bool askUserToAcceptWrongDigest(const zypp::Pathname &file,
+                                  const std::string &requested,
+                                  const std::string &found) override {
     if (callbacks == NULL || callbacks->checksum_wrong == NULL) {
-      return zypp::DigestReport::askUserToAcceptWrongDigest(file, requested, found);
+      return zypp::DigestReport::askUserToAcceptWrongDigest(file, requested,
+                                                            found);
     }
-    return callbacks->checksum_wrong(file.c_str(), requested.c_str(), found.c_str(), callbacks->checksum_wrong_data);
+    return callbacks->checksum_wrong(file.c_str(), requested.c_str(),
+                                     found.c_str(),
+                                     callbacks->checksum_wrong_data);
   }
 };
 
 static DigestReceive digest_receive;
 
+struct PatchScriptReport
+    : public zypp::callback::ReceiveReport<zypp::target::PatchScriptReport> {
+  struct InstallCallbacks *callbacks;
```

Comment:
Oh, I've missed this: we need at least a lifetime comment, better a smart pointer that provides some safety.

Reply:
well, smart pointer won't help you when rust deallocate it. Comment would be OK. Basically lifetime is for commit phase where we set and unset it ( it is same for all callbacks. I want to have it living only for limited time needed for operation ).